### PR TITLE
Add ws subscriber tests [newsHeads, logs]

### PR DESCRIPTION
**Pull request by lyubenkov**

### CODE REVIEW

File: deploy/requirements/prod.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-pytest==6.2.5
+pytest==7.4.0
```
Comment (gigimon) [inline deploy/requirements/prod.txt]: Did you check another tests with this pytest?

**Review by gigimon — COMMENTED** [review]
looks normal, but I don't like decision with classes for create requests

File: deploy/requirements/prod.txt (hunk 1 above)
Comment (lyubenkov) [reply deploy/requirements/prod.txt]: Reverted.

**Comment by lyubenkov** [comment]
> looks normal, but I don't like decision with classes for create requests

How would you rather do it?

File: integration/tests/basic/helpers/rpc_checks.py [hunk 2]
```diff
@@ -92,4 +92,7 @@ def assert_log_field_in_neon_trx_receipt(response, events_count):
 
 def assert_fields_are_hex(object, expected_hex_fields):
     for field in expected_hex_fields:
-        assert is_hex(object[field]), f"field {field} is not correct. Actual : {object[field]}"
+        try:
```
Comment (gigimon) [inline integration/tests/basic/helpers/rpc_checks.py]: Don't use try/ecept if you can rewrite this code on if/else. More useful like:
```
if field in object and not is_hex(object[field]): raise
if hasattr(field, object) ...
````

File: integration/tests/services/test_ws_subscriber.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+class Unsubscribe(ETH):
+    def __init__(self, subscription_id: str):
+        super().__init__([subscription_id], "eth_unsubscribe")
+
+
+@allure.feature("Websocket Subscriber")
+@allure.story("Subscribe to events")
+class TestSubscriber(TestRpcCalls):
+
+    async def test_eth_subscribe(self):
```
Comment (gigimon) [inline integration/tests/services/test_ws_subscriber.py]: test_eth_subsctibe -> test_subscribe_to_newHeads

File: integration/tests/services/test_ws_subscriber.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+            r = await ws.recv()
+            response = json.loads(r, object_hook=lambda d: SimpleNamespace(**d))
+            subscription = response.result
+            assert len(subscription) > 0
+
+            self.send_neon(self.sender_account, self.recipient_account, 0.1)
+
+            r = await ws.recv()
+            response = json.loads(r, object_hook=lambda d: SimpleNamespace(**d))
+            assert response.params.subscription == subscription
```
Comment (gigimon) [inline integration/tests/services/test_ws_subscriber.py:126]: add more checks for result field (https://ethereum.org/en/developers/tutorials/using-websockets/)

File: integration/tests/services/test_ws_subscriber.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+            assert len(vars(response.params.result).items()) > 0
+
+            while len(ws.messages) > 0:  # clear all received messages
+                ws.messages.popleft()
+
+            data = Unsubscribe(subscription)
+            await ws.send(json.dumps(data.__dict__))
+            r = await ws.recv()
+            response = json.loads(r, object_hook=lambda d: SimpleNamespace(**d))
+            assert response.result == 'true'
```
Comment (gigimon) [inline integration/tests/services/test_ws_subscriber.py:138]: check that you don't receive messages after unsubscribe

File: integration/tests/services/test_ws_subscriber.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+    @pytest.mark.parametrize("param_fields", [("address", "topics"), ("topics",), ("address",), ()], ids=str)
+    async def test_logs(self, param_fields):
+        async with websockets.connect(TEST_STAND) as ws:
+            event_caller, _ = self.web3_client.deploy_and_get_contract(
```
Comment (gigimon) [inline integration/tests/services/test_ws_subscriber.py]: you can deploy contract not inside async with, before connect to websockets and will be better if you deploy it once before this test and use in all parametrized tests

File: integration/tests/services/test_ws_subscriber.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+                        if topic in log.topics:
+                            is_arg_topic_in_list = True
+                else:
+                    is_arg_topic_in_list = True
+
+            assert (is_event_topic_in_list), f"Filter by {topics} works incorrect. Response: {m}"
+            assert (is_arg_topic_in_list), f"Filter by {topics} works incorrect. Response: {m}"
+
+    @pytest.mark.parametrize("subscription_type", ["newHeads", "logs"])
+    async def test_another_user_cant_unsubscribe(self, subscription_type: string):
```
Comment (gigimon) [inline integration/tests/services/test_ws_subscriber.py]: we need a test when a several users connect to one contract and subscribe to several event types and get only this

File: integration/tests/services/test_ws_subscriber.py [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+            tx = self.make_tx_object(self.sender_account.address)
+            instruction_tx = event_caller.functions.callEvent3(
+                arg1, arg2, arg3
+            ).build_transaction(tx)
+            self.web3_client.send_transaction(self.sender_account, instruction_tx)
+
+            messages = await ws_receive_all_messages(ws)
+            assert (len(messages) == log_count), f"Expected {log_count} event logs, but found {len(messages)}"
+
+            is_event_topic_in_list = False
```
Comment (gigimon) [inline integration/tests/services/test_ws_subscriber.py]: try to optimize this, maybe split it to several tests or something

**Review by gigimon — CHANGES_REQUESTED** [review]

File: integration/tests/services/test_ws_subscriber.py [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+
+
+class Unsubscribe(ETH):
+    def __init__(self, subscription_id: str):
+        super().__init__([subscription_id], "eth_unsubscribe")
+
+
+@allure.feature("Websocket Subscriber")
+@allure.story("Subscribe to events")
+class TestSubscriber(TestRpcCalls):
```
Comment (kristinaNikolaevaa) [inline integration/tests/services/test_ws_subscriber.py]: Why do you use the class with tests as a parent? TestRpcCalls

File: integration/tests/services/test_ws_subscriber.py [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+import allure
+import pytest
+import websockets
+from eth_utils import keccak
+
+from integration.tests.basic.helpers.rpc_checks import assert_fields_are_hex
+from integration.tests.basic.test_rpc_calls import TestRpcCalls
+from integration.tests.services.helpers.basic import ws_receive_all_messages
+
+TEST_STAND = "ws://159.69.194.181:8282"
```
Comment (kristinaNikolaevaa) [inline integration/tests/services/test_ws_subscriber.py]: Maybe it is better to move it to envs.json

File: integration/tests/basic/helpers/rpc_checks.py (hunk 2 above)
Comment (lyubenkov) [reply integration/tests/basic/helpers/rpc_checks.py]: Fixed

File: integration/tests/services/test_ws_subscriber.py (hunk 5 above)
Comment (lyubenkov) [reply integration/tests/services/test_ws_subscriber.py:138]: Fixed

File: integration/tests/services/test_ws_subscriber.py (hunk 3 above)
Comment (lyubenkov) [reply integration/tests/services/test_ws_subscriber.py]: Fixed

File: integration/tests/services/test_ws_subscriber.py (hunk 4 above)
Comment (lyubenkov) [reply integration/tests/services/test_ws_subscriber.py:126]: Fixed

File: integration/tests/services/test_ws_subscriber.py (hunk 6 above)
Comment (lyubenkov) [reply integration/tests/services/test_ws_subscriber.py]: Good idea, added fixture.

File: integration/tests/services/test_ws_subscriber.py (hunk 8 above)
Comment (lyubenkov) [reply integration/tests/services/test_ws_subscriber.py]: Fixed

File: integration/tests/services/test_ws_subscriber.py (hunk 7 above)
Comment (lyubenkov) [reply integration/tests/services/test_ws_subscriber.py]: Added

File: integration/tests/services/helpers/basic.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
+import asyncio
+import json
+import typing
+from types import SimpleNamespace
+
+from eth_utils import keccak
+from websockets.legacy.client import WebSocketClientProtocol
+
+
+async def ws_receive_all_messages(ws: WebSocketClientProtocol, timeout: int = 15) -> list:
```
Comment (gigimon) [inline integration/tests/services/helpers/basic.py]: move websocket specific functions to helpers/websockets. Mix async and sync functions in one modules is not a good practise

File: integration/tests/services/test_ws_subscriber.py [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+    def __init__(self, subscription_id: str):
+        super().__init__([subscription_id], "eth_unsubscribe")
+
+
+@allure.feature("Websocket Subscriber")
+@allure.story("Subscribe to events")
+class TestSubscriber(TestRpcCalls):
+
+    @pytest.fixture
+    def event_caller(self) -> typing.Any:
```
Comment (gigimon) [inline integration/tests/services/test_ws_subscriber.py]: event_caller -> event_caller_contract
In this case you will deploy new contract for each test

File: integration/tests/services/test_ws_subscriber.py [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+        )
+        yield event_caller
+
+    def call_contract_events(self, event_caller: typing.Any):
+        arg1, arg2, arg3 = ("text1", "text2", "text3")
+        tx = self.make_tx_object(self.sender_account.address)
+        instruction_tx = event_caller.functions.callEvent1(arg1).build_transaction(tx)
+        self.web3_client.send_transaction(self.sender_account, instruction_tx)
+
+        tx = self.make_tx_object(self.sender_account.address)
```
Comment (gigimon) [inline integration/tests/services/test_ws_subscriber.py]: why you don't use cycle here? like:
```
variants = {"callEvent1": [(arg1, ), (arg2, )], "callEvent2": [(arg2, arg3), (arg1, arg2)]}
for f in variants:
    for variant in variants[f]:
        tx = self.make_tx_object(self.sender_account.address)
        instruction_tx = getattr(event_caller.functions, f)(variant).build_transaction(
            tx
        )
        self.web3_client.send_transaction(self.sender_account, instruction_tx)

File: integration/tests/services/test_ws_subscriber.py [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+        instruction_tx = event_caller.functions.callEvent3(
+            arg1, arg2, arg3
+        ).build_transaction(tx)
+        self.web3_client.send_transaction(self.sender_account, instruction_tx)
+
+    @staticmethod
+    def assert_all_messages(messages: list, topics: list):
+        is_event_topic_in_list = False
```
Comment (gigimon) [inline integration/tests/services/test_ws_subscriber.py:77]: need documentation

File: integration/tests/services/test_ws_subscriber.py [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+            assert_fields_are_hex(response.params.result,
+                                  ["extraData", "gasLimit", "gasUsed", "logsBloom",
+                                   "nonce", "parentHash", "sha3Uncles",
+                                   "stateRoot", "timestamp", "transactionsRoot"])
+
+            data = Unsubscribe(subscription)
+            await ws.send(json.dumps(data.__dict__))
+            messages = await ws_receive_all_messages(ws)
+            response = messages[len(messages) - 1]
+            assert response.result == 'true'
```
Comment (gigimon) [inline integration/tests/services/test_ws_subscriber.py:138]: you don't check transactions inside newHeads and block number don't compare with block number from transaction receipt. But in real situation for small time you can get several blocks and you should check it in cycle and found head with your transaction and your block number
https://docs.alchemy.com/reference/newheads

**Review by gigimon — APPROVED** [review]

File: integration/tests/services/helpers/basic.py (hunk 11 above)
Comment (lyubenkov) [reply integration/tests/services/helpers/basic.py]: Moved

File: integration/tests/services/test_ws_subscriber.py (hunk 13 above)
Comment (lyubenkov) [reply integration/tests/services/test_ws_subscriber.py]: Great idea, improved!

File: integration/tests/services/test_ws_subscriber.py (hunk 14 above)
Comment (lyubenkov) [reply integration/tests/services/test_ws_subscriber.py:77]: Added

File: integration/tests/services/test_ws_subscriber.py (hunk 15 above)
Comment (lyubenkov) [reply integration/tests/services/test_ws_subscriber.py:138]: Added check by `blockHash`.

File: integration/tests/services/test_ws_subscriber.py (hunk 12 above)
Comment (lyubenkov) [reply integration/tests/services/test_ws_subscriber.py]: Renamed. Tests scope == 'class'

File: integration/tests/services/test_ws_subscriber.py (hunk 9 above)
Comment (lyubenkov) [reply integration/tests/services/test_ws_subscriber.py]: Outdated, it's `BaseMixin` now

File: integration/tests/services/test_ws_subscriber.py (hunk 10 above)
Comment (lyubenkov) [reply integration/tests/services/test_ws_subscriber.py]: Moved

**Review by kristinaNikolaevaa — APPROVED** [review]

File: .github/workflows/services.yml [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+          - 8
+          - 12
+          - auto
+env:
+  NETWORK: devnet
+  BUILD_URL: "${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}"
+  SOLANA_URL: "${{ secrets.SOLANA_URL }}"
+  NUMPROCESSES: 4
+  IMAGE:  neonlabsorg/neon_tests
+  CONTAINER:  basic-${{ github.run_id }}
```
Comment (kristinaNikolaevaa) [inline .github/workflows/services.yml]: not basic

File: .github/workflows/services.yml [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+      - name: "Generate allure report"
+        if: always()
+        id: requirements
+        uses: ./.github/actions/generate-allure-report
+        with:
+          container: ${{ env.CONTAINER }}
+          network: ${{ env.NETWORK }}
+          aws_access_key_id:  ${{ secrets.AWS_ACCESS_KEY_ID }}
+          aws_secret_access_key:  ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          tests_name: basic
```
Comment (kristinaNikolaevaa) [inline .github/workflows/services.yml]: not basic

File: .github/workflows/services.yml (hunk 16 above)
Comment (lyubenkov) [reply .github/workflows/services.yml]: Fixed, thanks!

File: .github/workflows/services.yml (hunk 17 above)
Comment (lyubenkov) [reply .github/workflows/services.yml]: Fixed, thanks!

**Review by kristinaNikolaevaa — APPROVED** [review]

**Pull request merged**